### PR TITLE
Feature/skeleton

### PR DESCRIPTION
**Pull request by jlconlin**

### CODE REVIEW

**Comment by apmccartney** [comment]
Summary of review:

SectionSkeleton
-----
+ remove MF and MAT member variables and getters except in the FIle skeleton and Material skeleton, respectively.
+ accept a head record as part of the ctor to populate any required values
+ when traversing the file, in addition to checking `tail.MT() == 0`, verify the iterator used for traversal is less than the end iterator
+ echo erroroneous lines when possible
+ Do not use the error logger when catching an exception thrown elsewhere
+ add bracket operator for accessing composed skeletons, e.g. 
```cpp
auto myMF3 = myMaterialSkeleton[3];
```

File: implementation/FileSkeleton/src/FileSkeleton.hpp.in [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  long& lineNumber ){
+    std::vector<SectionSkeleton<Iterator> > SI;
+
+    try{
+      SI.push_back(
+        SectionSkeleton<Iterator>(head, begin, position, end, lineNumber) );
+
+      auto FEND = StructureDivision(position, end, lineNumber); 
+
+      while( not FEND.isFend() ){
```
Comment (apmccartney) [inline implementation/FileSkeleton/src/FileSkeleton.hpp.in]: `FEND` (as a name) doesn't seem to convey the intent you had

File: implementation/FileSkeleton/src/FileSkeleton.hpp.in [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    throw e;
+  }
+
+template<typename Iterator>
+std::vector<SectionSkeleton<Iterator> > 
+FileSkeleton<Iterator>::createIndexedVector(
+  const HeadRecord& head,
+  Iterator begin, Iterator& position, const Iterator& end, 
+  long& lineNumber ){
+    std::vector<SectionSkeleton<Iterator> > SI;
```
Comment (apmccartney) [inline implementation/FileSkeleton/src/FileSkeleton.hpp.in]: Good hint for named return value optimization

**Comment by apmccartney** [comment]
In the `FileSkeleton` template, the `Iterator` template is forwarded to the `SectionSkeleton` template in numerous places. We could clean up the code a bit by declaring a using statement in the FileSkeleton body, e.g.
```cpp
template< typename Iterator >
class FileSkeleton {
public:
  using SectionSkeleton_t = SectionSkeleton< Iterator >; 
...
  SectionSkeleton_t
  operator[]( int MT );
...
};

File: implementation/FileSkeleton/src/FileSkeleton.hpp.in [hunk 3]
```diff
@@ -9,43 +9,68 @@ namespace implementation {
...
 
-    auto FEND = StructureDivision(position, end, lineNumber); 
+    try{
+      SI.push_back(
```
Comment (apmccartney) [inline implementation/FileSkeleton/src/FileSkeleton.hpp.in]: In this circumstance, it is safe to use emplace_back if you're interested, e.g.
```cpp
SI.emplace_back( head, begin, position, end, lineNumber );
```

File: implementation/FileSkeleton/src/FileSkeleton.hpp.in [hunk 4]
```diff
@@ -9,43 +9,68 @@ namespace implementation {
...
+    }
+
+    return SI;
+}
+
+template<typename Iterator>
+std::unordered_map<int, SectionSkeleton<Iterator>& > 
+FileSkeleton<Iterator>::createSectionsMap(){
+
+  std::unordered_map<int, SectionSkeleton<Iterator>& > sectionMap;
```
Comment (apmccartney) [inline implementation/FileSkeleton/src/FileSkeleton.hpp.in]: Again, good job with being mindful of rvo.

There's an opportunity to reserve the space you need at this point.

File: implementation/FileSkeleton.hpp.in [hunk 5]
```diff
@@ -11,27 +12,33 @@ class ENDFtk::implementation::FileSkeleton {
...
     Iterator begin_;  //!< The beginning of the FileSkeleton
-    Iterator end_;    //!< The end of the FileSkeleton
 
-    std::unordered_map<int, SectionSkeleton<Iterator> > sections_;
     std::vector<SectionSkeleton<Iterator> > sectionsIndexed_;
+    std::unordered_map<int, SectionSkeleton<Iterator>& > sections_;
+
+    Iterator end_;    //!< The end of the FileSkeleton
+
+    std::vector<SectionSkeleton<Iterator> >
```
Comment (apmccartney) [inline implementation/FileSkeleton.hpp.in]: this should be a static method if you want to safely call it in the initialization list

File: implementation/FileSkeleton.hpp.in [hunk 6]
```diff
@@ -11,27 +12,33 @@ class ENDFtk::implementation::FileSkeleton {
...
+
+    Iterator end_;    //!< The end of the FileSkeleton
+
+    std::vector<SectionSkeleton<Iterator> > 
+    createIndexedVector(
+      const HeadRecord& head,
+      Iterator begin, Iterator& position, const Iterator& end, 
+      long& lineNumber );
+
+    std::unordered_map<int, SectionSkeleton<Iterator>& >
```
Comment (apmccartney) [inline implementation/FileSkeleton.hpp.in]: this should be a static method if you want to safely call it in the initialization list.

Accept a `const std::vector< SectionSkeleton<Iterator> >&` as an argument here, passing the member variable during invocation in the initialization list

File: implementation/FileSkeleton/src/FileSkeleton.hpp.in [hunk 7]
```diff
@@ -9,43 +9,68 @@ namespace implementation {
...
+
+  try{
+    for( auto& sec : sectionsIndexed_ ){
```
Comment (apmccartney) [inline implementation/FileSkeleton/src/FileSkeleton.hpp.in]: good job avoiding copies

File: implementation/FileSkeleton/src/FileSkeleton.hpp.in [hunk 8]
```diff
@@ -9,43 +9,68 @@ namespace implementation {
...
+        SectionSkeleton<Iterator>(head, begin, position, end, lineNumber) );
 
-    while( not FEND.isFend() ){
-      LOG(INFO) << "diff before: " << end-position;
-      if(position >= end){
-        LOG(ERROR) << "File encountered end of stream before reading FEND";
-        throw std::exception();
+      auto FEND = StructureDivision(position, end, lineNumber); 
+
+      while( not FEND.isFend() ){
```
Comment (apmccartney) [inline implementation/FileSkeleton/src/FileSkeleton.hpp.in]: while FEND is not a fend...

The name choice `FEND` might not be conveying your intent here

File: .travis.sh [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    sudo update-alternatives --auto gcov
+    export SPECIFY_LD="-fuse-ld=gold -fuse-linker-plugin"
+    if [ $CXX = "g++" ]
+    then
+        export CC="/usr/bin/gcc-6"
+        export CXX="/usr/bin/g++-6"
+        export AR="/usr/bin/gcc-ar-6"
+        export RANLIB="/usr/bin/gcc-ranlib-6"
+        export NM="/usr/bin/gcc-nm-6"
+        export DEPTH="2048"
```
Comment (jlconlin) [inline .travis.sh]: <grumble> gcc </grumble>

**Pull request merged**